### PR TITLE
Fix for SI-6731, dropped trees in selectDynamic.

### PR DESCRIPTION
**Pull request by paulp**
I rewrote mkInvoke entirely, and boosted the test coverage.

### CODE REVIEW

**Comment by jannic** [comment]
Looks much better than the band-aid solution I submitted in https://github.com/scala/scala/pull/1621. That's exactly what I meant when I asked "Is there a better way to fix this than to add all such cases manually to mkInvoke?".

File: test/files/pos/t6720.scala [hunk 1]
```diff
@@ -0,0 +1,11 @@
```
Comment (adriaanm) [inline test/files/pos/t6720.scala]: I believe this is SI-6722, please rename and push -f

Comment (adriaanm) [inline test/files/pos/t6720.scala]: (And update commit message)

**Comment by adriaanm** [comment]
Sorry, I thought about this some more and I think this should go into 2.10.1, as did #1621.

**Comment by paulp** [comment]
Okay... and the PR was opened against 2.10.x. Did you close it for some reason which I should be able to discern?

**Comment by paulp** [comment]
I'm going to reopen this so I don't forget about it, in the belief you closed this thinking it was opened against -wip. If I misunderstood something, please disabuse me of my abused notions.

**Comment by adriaanm** [comment]
my bad

**Comment by paulp** [comment]
Review by @adriaanm the rogue closer.

**Comment by retronym** [comment]
lgtm, way to kill N birds with one stone.

**Comment by retronym** [comment]
@paulp: just to confirm, this closes:

SI-6731, SI-5726, SI-5733, SI-6320, SI-6551 and SI-6722. 

Any others?

**Comment by paulp** [comment]
There was one other which was already closed after a partial fix covered its example. So, no others.

**Comment by adriaanm** [comment]
(ah, this is where I meant to type my thanks cleaner-upper @paulp comment, and again, @retronym for closing the tickets i missed -- i'll crawl back under my rock now)

**Pull request merged**